### PR TITLE
Trust native notebooks using VSC API

### PR DESCRIPTION
**Pull request by DonJayamanne**

### CODE REVIEW

File: src/client/datascience/notebook/helpers/helpers.ts [hunk 1]
```diff
@@ -275,9 +276,9 @@ function createCodeCellFromNotebookCell(cell: NotebookCell): nbformat.ICodeCell
     };
 }
 
-function createNotebookCellDataFromRawCell(isNbTrusted: boolean, cell: nbformat.IRawCell): NotebookCellData {
+function createNotebookCellDataFromRawCell(cell: nbformat.IRawCell): NotebookCellData {
     const notebookCellMetadata = new NotebookCellMetadata().with({
-        editable: isNbTrusted,
+        editable: true,
```
Comment (DonJayamanne) [inline src/client/datascience/notebook/helpers/helpers.ts:281]: When the notebook is not editable, then cells are not editable.
This way we don't need to update cell metadata to indicate its editable or not.
Leaving it as always editable is easier (just toggle at the document metatadata level).

File: src/client/datascience/notebook/helpers/helpers.ts [hunk 2]
```diff
@@ -186,6 +186,7 @@ export function notebookModelToVSCNotebookData(
             editable: isNotebookTrusted,
```
Comment (DonJayamanne) [inline src/client/datascience/notebook/helpers/helpers.ts:186]: Notebook level metadata, this overrides what's at each cell level.

File: src/client/datascience/notebook/helpers/helpers.ts [hunk 3]
```diff
@@ -186,6 +186,7 @@ export function notebookModelToVSCNotebookData(
...
             cellHasExecutionOrder: true,
             runnable: isNotebookTrusted,
+            trusted: isNotebookTrusted,
```
Comment (DonJayamanne) [inline src/client/datascience/notebook/helpers/helpers.ts:189]: New VSC API

File: src/client/datascience/notebookStorage/vscNotebookModel.ts [hunk 4]
```diff
@@ -126,28 +117,30 @@ export class VSCodeNotebookModel extends BaseNotebookModel {
...
-            if (editor) {
-                await updateVSCNotebookAfterTrustingNotebook(editor, this.document, this._cells);
-            }
-            // We don't need old cells.
-            this._cells = [];
-            this._trustedAfterOpeningNotebook = true;
+        this.trust();
+        const editor = this.vscodeNotebook?.notebookEditors.find((item) => item.document === this.document);
+        const document = editor?.document;
+        if (editor && document && !document.metadata.trusted) {
```
Comment (DonJayamanne) [inline src/client/datascience/notebookStorage/vscNotebookModel.ts:124]: When trusting all we now need to do is, just update the notebook document meteadata

File: src/client/datascience/interactive-ipynb/trustCommandHandler.ts [hunk 5]
```diff
@@ -90,42 +86,17 @@ export class TrustCommandHandler implements IExtensionSingleActivationService {
...
-            ]);
-            sendTelemetryEvent(Telemetry.TrustNotebook);
-        } finally {
-            trustedNotebookInTrustService.resolve();
-        }
+        // Update model trust
+        await model.trustNotebook();
+        // Trust the original contents & contents generated vy using the VSC Notebook model.
+        // When generating JSON from VSC, sometimes contents can get auto formatted based on user settings (hence that JSON in ipynb could be different from original contents).
+        // Thus trust both, the original content & the new content generated by VSC.
```
Comment (IanMatthewHuff) [inline src/client/datascience/interactive-ipynb/trustCommandHandler.ts:93]: Code looks good. But I just want to be careful about this here. I understand the comment here, but technically this is trusting both the original disk content and the content as formatted by VSC. The formatting done by VSC is something that is not specifically in our control. If we trust both A.ipynb and A.ipynb as formatted by vs code are we possibly also trusting B.ipynb where B.ipynb on disk is the same as A when formatted by VS Code?

If formatting touched no real content this wouldn't be an issue but I'm not sure if we have that guarantee.

**Review by rchiodo — APPROVED** [review]

**Review by DavidKutu — APPROVED** [review]

File: src/client/datascience/interactive-ipynb/trustCommandHandler.ts (hunk 5 above)
Comment (DonJayamanne) [reply src/client/datascience/interactive-ipynb/trustCommandHandler.ts:93]: Yes, that's correct we trust both formatted & unformatted models.
The trust state of A.ipynb & B.ipynb are not stored together, they are always different (trust state is keyed off the file Uri).

**Review by IanMatthewHuff — APPROVED** [review]

**Pull request merged**